lib/ide/src/controller/project.rs

```diff
@@ -78,47 +76,183 @@ impl Handle {
+            None => {
+                let loaded = self.create_text_controller(path.clone()).await?;
+                //TODO[ao] Here we should make a better solution for case where we simultaneously
+                // load one module twice.
+                // This is duplicated with open_module, but it's not worth refactoring as its
+                // temporary solution.
+                let cached = self.with_borrowed(|data|
+                    match data.text_cache.entry(path) {
+                        Occupied(entry) => entry.get().clone_ref(),
+                        Vacant(entry)   => entry.insert(loaded)
+                    }
+                );
+                Ok(cached)
+            }
+        }
+    }
+
+    /// Returns a module controller which have module opened from file.
+    pub async fn open_module(&self, loc:ModuleLocation)
+    -> FallibleResult<controller::module::Handle> {
+        let cached = self.with_borrowed(|data| data.module_cache.get(&loc));
+        match cached {
+            Some(controller) => Ok(controller),
+            None => {
+                let loaded = self.create_module_controller(loc.clone()).await?;
+                //TODO[ao] Here we should make a better solution for case where we simultaneously
+                // load one module twice.
+                let cached = self.with_borrowed(|data|
+                    match data.module_cache.entry(loc) {
+                        Occupied(entry) => entry.get().clone_ref(),
+                        Vacant(entry)   => entry.insert(loaded)
+                    }
+                );
+                Ok(cached)
+            },
+        }
+    }
+
+    async fn create_text_controller(&self, path:fmc::Path)
+    -> FallibleResult<controller::text::Handle> {
+        match ModuleLocation::from_path(&path) {
+            Some(location) => {
+                let module = self.open_module(location).await?;
+                Ok(controller::text::Handle::new_for_module(module))
+            },
+            None => {
+                let fm = self.file_manager();
+                Ok(controller::text::Handle::new_for_plain_test(path,fm))
+            }
+        }
+    }
+
+    async fn create_module_controller(&self, location:ModuleLocation)
+    -> FallibleResult<controller::module::Handle> {
+        let (fm,parser) = self.with_borrowed(|d| (d.file_manager.clone_ref(),d.parser.clone_ref()));
+        controller::module::Handle::new(location,fm,parser).await
+    }
 }
 
+
+
 #[cfg(test)]
 mod test {
     use super::*;
 
+    use crate::executor::global::spawn;
+    use crate::executor::global::set_spawner;
+
     use file_manager_client::Path;
     use json_rpc::test_util::transport::mock::MockTransport;
+    use futures::executor::LocalPool;
+    use wasm_bindgen_test::wasm_bindgen_test;
+    use wasm_bindgen_test::wasm_bindgen_test_configure;
+
+    wasm_bindgen_test_configure!(run_in_browser);
+
+
 
-    #[test]
+    #[wasm_bindgen_test]
```

Comment:
As previously and below. For debugging I'd really want to keep ability to run natively as many tests as possible.

Reply:
I agree. But I also would not make a fixture running scala parser as a part of this already big task.